plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/storage/LinstorStorageAdaptor.java

```diff
@@ -0,0 +1,566 @@
+        if (!MapStorageUuidToStoragePool.isEmpty())
+        {
+            String firstKey = MapStorageUuidToStoragePool.keySet().stream().findFirst().get();
+            final KVMStoragePool pool = MapStorageUuidToStoragePool.get(firstKey);
+
+            s_logger.debug("Linstor: Using storpool: " + pool.getUuid());
+            final DevelopersApi api = getLinstorAPI(pool);
+
+            try
+            {
+                List<ResourceWithVolumes> resources = api.viewResources(
+                    Collections.singletonList(getHostname()),
+                    null,
+                    null,
+                    null,
+                    null,
+                    null);
+
+                Optional<ResourceWithVolumes> rsc = getResourceByPath(resources, localPath);
+
+                if (rsc.isPresent())
+                {
+                    ResourceDefinitionModify rdm = new ResourceDefinitionModify();
+                    rdm.deleteProps(Collections.singletonList("DrbdOptions/Net/allow-two-primaries"));
+                    ApiCallRcList answers = api.resourceDefinitionModify(rsc.get().getName(), rdm);
+                    if (ApiCallChecker.hasError(answers))
+                    {
+                        s_logger.error("Failed to remove 'allow-two-primaries' on " + rsc.get().getName());
+                        throw new CloudRuntimeException(answers.get(0).getMessage());
+                    }
+
+                    return true;
+                }
+                s_logger.warn("Linstor: Couldn't find resource for this path: " + localPath);
+            } catch (ApiException apiExc) {
+                s_logger.error(apiExc);
+                throw new CloudRuntimeException(apiExc.getMessage());
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean deletePhysicalDisk(String name, KVMStoragePool pool, Storage.ImageFormat format)
+    {
+        s_logger.debug("Linstor: deletePhysicalDisk " + name);
+        final DevelopersApi api = getLinstorAPI(pool);
+
+        try {
+            final String rscName = getLinstorRscName(name);
+            s_logger.debug("Linstor: delete resource definition " + rscName);
+            ApiCallRcList answers = api.resourceDefinitionDelete(rscName);
+            handleLinstorApiAnswers(answers, "Linstor: Unable to delete resource definition " + rscName);
+        } catch (ApiException apiExc) {
+            throw new CloudRuntimeException(apiExc);
+        }
+        return true;
+    }
+
+    @Override
+    public KVMPhysicalDisk createDiskFromTemplate(
+        KVMPhysicalDisk template,
+        String name,
+        QemuImg.PhysicalDiskFormat format,
+        Storage.ProvisioningType provisioningType,
+        long size,
+        KVMStoragePool destPool,
+        int timeout)
+    {
+        s_logger.info("Linstor: createDiskFromTemplate");
+        // if source is linstor lvm-thin snapshot we could snapshot and restore from, but do plain copy for now
```

Comment:
@rp- I do not know Linstor and do not know why you choose to use this way, but I am wondering why do a plain copy instead of a snapshot...

Reply:
@rp- ok, understood.

I know from `qemu` that there is a way to coalesce (merge) or collapse the snapshots ([cheatsheet](https://blog.programster.org/qemu-img-cheatsheet)). Could ZFS have anything like that?